### PR TITLE
Allow lambda in menu item :match_path option and URL

### PR DESCRIPTION
**Pull request by mamhoff**
## Summary

This allows setting a path for a menu item via a lambda executed at runtime. If engines other than spree would like their route in the main menu, this is how they can do dynamically without having to hard-code the path. 

Similarly, we allow the `match_path` option for Menu Items to be a callable taking a request. 

## Checklist

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- ✅ I have added automated tests to cover my changes.

### CODE REVIEW

File: backend/lib/spree/backend_configuration/menu_item.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+require 'spree/preferences/configuration'
+require 'spree/backend_configuration/menu_item'
```
Comment (tvdeyen) [inline backend/lib/spree/backend_configuration/menu_item.rb:4]: These requires should not be necessary

**Review by tvdeyen — APPROVED** [review]
Nice feature. This helps engines like [alchemy-solidus](https://github.com/AlchemyCMS/alchemy-solidus) to hook into the main navigation much nicer 👍🏻 

_Rubocop still needs some attention here._

**Comment by mamhoff** [comment]
Rubocop issues should be addressed.

File: backend/lib/spree/backend_configuration/menu_item.rb (hunk 1 above)
Comment (elia) [reply backend/lib/spree/backend_configuration/menu_item.rb:4]: `require 'spree/backend_configuration/menu_item'` isn't this a circular require?

Ruby handles it fine, but we should remove just that line, the other one is a direct dependency and it's ok to keep it there 👍

**Review by elia — COMMENTED** [review]
I like this change a lot, I think this is the right direction 🙌 

I'm also wondering if we can also integrating some of the logic in `backend/app/views/spree/admin/shared/_tabs.html.erb`.

Maybe we want to evaluate the url in the context of the view like we do for the `condition`: `instance_exec(&menu_item.condition)`.

And we might also think of ways to move this ternary from the view to the `#url` helper `menu_item.url.is_a?(Symbol) ? spree.public_send(menu_item.url) : menu_item.url`.

@mamhoff thoughts on this kind of changes?

File: backend/lib/spree/backend_configuration/menu_item.rb (hunk 1 above)
Comment (tvdeyen) [reply backend/lib/spree/backend_configuration/menu_item.rb:4]: > isn't this a circular require?

Yep, thats what I thought as well

Comment (mamhoff) [reply backend/lib/spree/backend_configuration/menu_item.rb]: Both are unnecessary as the file is required from a file that has both. Removed, thank you Thomas, and sorry for the misunderstanding

Comment (tvdeyen) [reply backend/lib/spree/backend_configuration/menu_item.rb]: No worries

**Comment by mamhoff** [comment]
While I think this API reorganization would be a nice idea, I don't think the level of effort and change involved is worth the pain at this point...

**Review by jarednorman — APPROVED** [review]

**Review by elia — APPROVED** [review]
Fair enough, that's admittedly a lot 👍👍👍

**Pull request merged**